gateway/gateway.go

```diff
@@ -670,17 +678,40 @@ func (gateway *HandleT) getJobDataFromRequest(req *webRequestT) (jobData *jobFro
 			`{"error": "rudder-server gateway failed to marshal params"}`,
 		)
 	}
-	err = nil
-	job := &jobsdb.JobT{
-		UUID:         id,
-		UserID:       firstUserID,
-		Parameters:   marshalledParams,
-		CustomVal:    CustomVal,
-		EventPayload: body,
-		EventCount:   jobData.numEvents,
-		WorkspaceId:  workspaceId,
-	}
-	jobData.job = job
+	jobs := make([]*jobsdb.JobT, len(out))
+	for idx, event := range out {
+		var payload json.RawMessage
+		{
+			type SingularEventBatch struct {
+				Batch      []types.SingularEventT `json:"batch"`
+				RequestIp  string                 `json:"requestIP"`
+				WriteKey   string                 `json:"writeKey"`
+				ReceivedAt string                 `json:"receivedAt"`
+			}
+			singularEventBatch := SingularEventBatch{
+				Batch:      []types.SingularEventT{event},
+				RequestIp:  ipAddr,
+				WriteKey:   writeKey,
+				ReceivedAt: time.Now().Format(misc.RFC3339Milli),
+			}
+			payload, err = json.Marshal(singularEventBatch)
+			if err != nil {
+				err = errors.New(response.InvalidJSON)
+				return
+			}
+		}
+
+		jobs[idx] = &jobsdb.JobT{
+			UUID:         uuid.New(),
+			UserID:       firstUserID,
+			Parameters:   marshalledParams,
+			CustomVal:    CustomVal,
+			EventPayload: payload,
+			EventCount:   1,
+			WorkspaceId:  workspaceId,
```

Comment:
Should we also populate the payload size field?
```suggestion
			EventCount:   1,
			PayloadSize: len(payload),
			WorkspaceId:  workspaceId,
```

Also, we might want to populate the userID of the payload instead of `firstUserID`. Although it breaks compatibility with our current approach and might introduce performance downsides.
.

Reply:
True
We consider each individual event's userID for suppression purposes anyway.
Elaborate further on the performance downsides.. I don't see a drastic effect due to this at all.

Reply:
adopted👍